ui: Fix the custom time picker reverting to the previously selected time, part 2

jocrl opened the pull request:
Fixes a bug was introduced in #80660, where if a user selects a different time in the
custom time selection, the component may revert to the currently select
time. #81006 had fixed this bug for when a custom time was selected. However,
it had not fixed the bug for when a non-custom time, e.g. "Past 10 minutes", was
selected.

This commit fixes the bug for when a non-custom time is selected. The fix
implicitly relies on the (currently true) assumption that a user-induced time
change always closes the dropdown (sets `isVisible` to false) or occurs when
the dropdown is already closed (`isVisible` to already false).

Before, the bug is present when "Past 10 minutes" is selected. No bug when a custom time is selected:
https://user-images.githubusercontent.com/91907326/167907768-59f98662-8ca4-4ea3-ad2e-69e90e252800.mov

After, Statements Page: 
https://user-images.githubusercontent.com/91907326/167908454-1d5508ca-399b-4c39-9fdb-7362e06cc0f1.mov

After, Metrics Page: 
https://user-images.githubusercontent.com/91907326/167910508-2624c982-19b0-4b29-bc43-9518d055ecc2.mov

However, this implementation comes at the cost that temporary, un-applied
custom user selections will be lost when the user closes the custom picker.

Before, temporary, un-applied custom user selections (i.e. without clicking "Apply") would persist between show/hides of the picker:
https://user-images.githubusercontent.com/91907326/167913174-aa1b7478-2711-46a7-b837-b44feac2c9ef.mov

After, the cost of this fix's implementation is that un-applied custom selections no longer persist:
https://user-images.githubusercontent.com/91907326/167913291-b12af862-8f90-431b-9f18-695d12f440cb.mov


To explain motivation for the existing code, the `useEffect` was previously needed in the absence of triggering component re-mounting through `isVisible`. If one removed the `useEffect`s without adding the `isVisible` re-mounting trigger, the initial time would not update when the Dropdown visibility was changed (because the visibility change is implemented via a css change):

https://user-images.githubusercontent.com/91907326/167908109-0b1bc17d-43b8-45f0-ab0a-0b07ca33cb22.mov


Release note: None

Release justification: Category 2, UI bug fix

jocrl committed:
ui: Fix the custom time picker reverting to the previously selected time, pt 2

Fixes a bug was introduced in #80660, where if a user selects a different time in the
custom time selection, the component may revert to the currently select
time. #81006 had fixed this bug for when a custom time was selected. However,
it had not fixed the bug for when a non-custom time, e.g. "Past 10 minutes", was
selected.

This commit fixes the bug for when a non-custom time is selected. The fix
implicitly relies on the (currently true) assumption that a user-induced time
change always closes the dropdown (sets `isVisible` to false) or occurs when
the dropdown is already closed (`isVisible` to already false).

However, this implementation comes at the cost that temporary, un-applied
custom user selections will be lost when the user closes the custom picker.

Release note: None

Release justification: Category 2, UI bug fix

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81196)
<!-- Reviewable:end -->

jocrl commented:
Hello reviewers! So, https://github.com/cockroachdb/cockroach/pull/81006 only partially fixed a bug I had introduced in https://github.com/cockroachdb/cockroach/pull/80660. It fixed the bug when a custom time was selected, but left a bug remainig for when time relative to now was selected, e.g. "Past 10 minutes".

I considered two approaches to fixing the "Past 10 minutes" situation:
1) This commit's implementation
2) If there was a stable, fixed notion of "now" that updated only when the page polled and the data updated (rather than during arbitrary component renders), I could use that. However, for the Statements and Transactions page the polling time is tied up in an implicit throttle of CachedDataReducer. Polling for the DB Console Metrics page is also gnarly.

Option 2) sounds expensive, or blocked on cleaning up polling. So I went with 1), which comes with the cost described in the commit message.

matthewtodd reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81196#-:-N1xxya78SDngs0Aa0mC:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81196#-:-N1yf696-KjSQOAqFM8n:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @jocrl)



<!-- Sent from Reviewable.io -->

jocrl commented:
bors r-

jocrl commented:
bors r+
